crates/ruff_linter/src/rules/pylint/rules/bad_open_mode.rs

```diff
@@ -85,23 +85,20 @@ enum Kind {
 
 /// If a function is a call to `open`, returns the kind of `open` call.
 fn is_open(func: &Expr, semantic: &SemanticModel) -> Option<Kind> {
-    match func {
-        // Ex) `pathlib.Path(...).open(...)`
-        Expr::Attribute(ast::ExprAttribute { attr, value, .. }) if attr.as_str() == "open" => {
-            match value.as_ref() {
-                Expr::Call(ast::ExprCall { func, .. }) => semantic
-                    .resolve_qualified_name(func)
-                    .is_some_and(|qualified_name| {
-                        matches!(qualified_name.segments(), ["pathlib", "Path"])
-                    })
-                    .then_some(Kind::Pathlib),
-                _ => None,
-            }
-        }
-        // Ex) `open(...)`
-        Expr::Name(ast::ExprName { id, .. }) => {
-            (id.as_str() == "open" && semantic.is_builtin("open")).then_some(Kind::Builtin)
-        }
+    // Ex) `open(...)`
+    if semantic.match_builtin_expr(func, "open") {
+        return Some(Kind::Builtin);
+    }
+
+    // Ex) `pathlib.Path(...).open(...)`
+    let ast::ExprAttribute { attr, value, .. } = func.as_attribute_expr()?;
+    if attr != "open" {
+        return None;
+    }
+    let ast::ExprCall { func, .. } = value.as_call_expr()?;
+    let qualified_name = semantic.resolve_qualified_name(func)?;
```

Comment:
Nit: This code confused me a lot because the line above rebinds `func` which I assumed was the `open` function. It might be worth giving it another name.